### ISSUE TITLE
saw-script interpreter typechecks one line at a time

### ISSUE THREAD

**Issue by sauclovian-g** [GaloisInc/saw-script#2158]
The saw-script interpreter typechecks one statement at a time when executing at the top level (either the TopLevel top level or the ProofScript top level, I think) ... and this is not easily fixed because it does special things with top-level binds that ought to be happening in the typechecker instead, except that the AST representation can't represent them so something needs to give.

It also just doesn't check some statements before executing them.

This is all made worse by the corresponding assumption in the typechecker that statements happen only inside do-blocks.

With #2157 the overtly incorrect behavior where top-level typedefs were just not checked at all has been corrected, but because of these complications that was done as a patch rather than a fix.

**Comment by sauclovian-g** [comment]
There are additional problems, such as that it turns out the type signatures in the list of builtin primitives are not checked. (Accordingly, at least one of them is wrong.) I think this needs to be fixed so that parsed but not typechecked text is not available to the interpreter code.

**Comment by RyanGlScott** [comment]
Some other unfortunate side effects of https://github.com/GaloisInc/saw-script/issues/2158#issuecomment-2897543562:

* The Yosys-related documentation in https://galoisinc.github.io/saw-script/master/saw-user-manual/analyzing-hardware-circuits-using-yosys.html no longer typechecks, as running `type m` on its own at the REPL will fail to typecheck.
* It is no longer as straightforward to print auxiliary theorems in `proof_subshell`. For instance, given the following SAWScript file:

  ```
  enable_experimental;

  let verify_xor_involutes (a: Type) : TopLevel Theorem = prove_print rme {{ \(x: a) -> \(y: a) -> (x ^ y) ^ y == x }};

  let verify_xor_involutes' (a: Type) : TopLevel Theorem = do {
      xor_involutes_lemma <- verify_xor_involutes a;

      prove_print do {
          proof_subshell ();
      } {{ \(x: a) -> \(y: a) -> \(z: a) -> (((x ^ y) ^ y) ^ z) ^ z == x }};
  };

  verify_xor_involutes' {| [128] |};
  ```

  With SAW 1.2, one can easily inspect `xor_involutes_lemma` in a proof subshell:

  ```
  $ ~/Software/saw-1.2/bin/saw test.saw
  [11:11:17.933] Loading file "/Users/rscott/Documents/Hacking/SAW/test.saw"
  proof (1)> xor_involutes_lemma
  [11:11:22.278] Theorem (let { x@1 = TCNum 128
        x@2 = Vec 128 Bool
        x@3 = PLogicSeqBool x@1
      }
   in (x : x@2)
  -> (y : x@2)
  -> EqTrue
       (ecEq x@2 (PEqSeqBool x@1) (ecXor x@2 x@3 (ecXor x@2 x@3 x y) y)
          x))
  ```

  With SAW 1.3, however, this same interaction fails to typecheck:

  ```
  $ ~/Software/saw-1.3/bin/saw test.saw
  [11:12:05.652] Loading file "/Users/rscott/Documents/Hacking/SAW/test.saw"
  proof (1)> xor_involutes_lemma

  <stdin>:1:1-1:20: Type mismatch.
      Mismatch of types.
      <stdin>:1:1-1:20: The type ProofScript t.0 arises from this type annotation
      <type of prove_print>:1:36-1:43: The type Theorem arises from this type annotation

      Expected: ProofScript t.0
      Found:    Theorem

      within "<proofscript>" (<stdin>:1:1-1:20)
  ```

  Note that because `proof_subshell` uses `ProofScript`, not `TopLevel`, we cannot even `print` the lemma! As such, the only way to inspect it is to `return` it:

  ```
  proof (1)> return xor_involutes_lemma
  [11:13:11.012] Theorem (let { x@1 = TCNum 128
        x@2 = Vec 128 Bool
        x@3 = PLogicSeqBool x@1
      }
   in (x : x@2)
  -> (y : x@2)
  -> EqTrue
       (ecEq x@2 (PEqSeqBool x@1) (ecXor x@2 x@3 (ecXor x@2 x@3 x y) y)
          x))
  ```

  Quite annoying.

**Comment by sauclovian-g** [comment]
Yeah, this got documented in the changelog but not properly anywhere in the issue tracker.

Also, while we're at it, we should change the type of print from `{a} a -> TopLevel ()` to `{m a} a -> m ()` so it'll run in any of the monads.
